tools/dev-server/index.bzl

```diff
@@ -0,0 +1,125 @@
+load("@build_bazel_rules_nodejs//internal/common:sources_aspect.bzl", "sources_aspect")
+
+"""
+  Gets the workspace name of the given rule context.
+"""
+
+def _get_workspace_name(ctx):
+    if ctx.label.workspace_root:
+        # We need the workspace_name for the target being visited.
+        # Skylark doesn't have this - instead they have a workspace_root
+        # which looks like "external/repo_name" - so grab the second path segment.
+        return ctx.label.workspace_root.split("/")[1]
+    else:
+        return ctx.workspace_name
+
+"""
+  Implementation of the dev server rule.
+"""
+
+def _dev_server_rule_impl(ctx):
+    files = depset(ctx.files.srcs)
+    required_tools = ctx.files._bash_runfile_helpers + \
+                     ctx.files._dev_server_bin + \
+                     ctx.attr._dev_server_bin[DefaultInfo].files.to_list() + \
+                     ctx.attr._dev_server_bin[DefaultInfo].data_runfiles.files.to_list()
+
+    for d in ctx.attr.deps:
+        if hasattr(d, "node_sources"):
+            files = depset(transitive = [files, d.node_sources])
+        elif hasattr(d, "files"):
+            files = depset(transitive = [files, d.files])
+        if hasattr(d, "dev_scripts"):
+            files = depset(transitive = [files, d.dev_scripts])
+
+    workspace_name = _get_workspace_name(ctx)
+    root_paths = ["", "/".join([workspace_name, ctx.label.package])] + ctx.attr.additional_root_paths
+
+    args = '--root_paths="%s" ' % ",".join(root_paths)
```

Comment:
Any reason not to use `ctx.actions.args()`? Similar to our `sass_bundle.bzl`

Reply:
The reason is that `ctx.actions.args()` is for Bazel actions. Since we don't execute an action, but rather create an executable, we can't use the `args` object. There is no public API to convert that into a string (i looked but didn't find anything; happy to try more if you have some ideas)

Reply:
Added a comment.